Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+
+### User story 2: Cloud provider implementing SCS stack
+
+After the cloud provider implemented the SCS cluster stack, all applications are ready to be installed and work perfectly. Therefore, the cloud provider does not have to invest time in configuring applications but can rely on the pre-defined and well-maintained SCS stack. Its users can profit from the additional value generated by this feature of ready-to-install applications. E2E tests ensure that everything has been implemented correctly.
+
+### User story 3: User wanting to implement a multi-cloud
+
+As the applications can be used on every SCS-certified cloud, a true multi-cloud is possible.
+
+### User story 4: SCS operator wanting to have maintainable SCS infrastructure
+
+Through few but well-defined cluster stacks, most use-cases of cloud providers and users are covered. These few stacks can be easily maintained by the SCS operator. Automated E2E tests reduce the manual work.
+
+## Risks and Mitigations
+
+The risks are the same as every standardization has. Cloud providers might not be satisfied by the selection of cluster stacks that SCS chose. As soon as one wish is granted, others come and want to have their ideas implemented in SCS as well. In the end, there is no standardization and a useless certification of SCS infrastructure.
+
+## Design details
+
+As described above, we want to achieve two things: standardizing clusters and making applications independent of the cluster infrastructure itself. By separating these two layers, we achieve a clear definition of responsibilities and compatibility of different cloud providers.
+
+In this way, standardized but also configurable Kubernetes clusters can be created in self-service - also via GitOps.
+
+### Proposal 1: Using cluster stacks
+
+There are various components playing a role in defining cluster infrastructure. We consider the following three technical layers:
+
+1. Pre-cluster state - node image
+2. Cluster state - Cluster API and provider implementation
+3. Kubernetes core applications
+
+All of these layers depend on each other, such that it is not possible to freely select and exchange components. This is why we introduce the cluster stack which consists of a selection of sub-systems and configurations that follow a certain goal.
+
+For example, there can be a stack for Kubernetes version 1.22 and one for 1.23. There can also be stacks that use a node image based on Ubuntu and others with Fedora. These cluster stacks have to be implemented by the providers, as there are certain details, e.g. related to the node image, that have to be adapted to the respective provider.
```

Comment:
Talking about the node image: Do we pin down the distribution which needs to be used?

I'm asking because RHEL/CentOS has some funny defaults which [*do* leak down into the container abstraction](https://gitlab.com/yaook/operator/-/issues/121#note_543076539) and hunting all of those down and pulling them straight sounds like a pretty daunting task. (And leading to fun hacks like [`realtime-hack.yaml`](https://gitlab.com/yaook/operator/-/blob/devel/ci/devel_integration_tests/deploy/realtime-hack.yaml#L20))

(FWIW, you could also read it the other way: Ubuntu has defaults which allow things which are (and possibly should not be) not allowed on RHEL and that affects portability.)

Also, SELinux and AppArmor also leak into k8s, keyword `securityContext`.

Reply:
The point I was trying to make is that attempting to specify what a distribution must deliver is going to be impossible to do completely. I never knew about that realtime scheduling parameter until the different default on CentOS broke stuff.

So if we care about portability, this is a challenge we must address in some way.

Reply:
There are also good reasons to derive from popular distributions (centos, ubuntu) to something build for light container runtimes like flatcar, coreos, and I bet there are many more.

For me, defining a distribution is again an implementation detail that should not be defined.

Reply:
@horazont I completely agree with you. This is why we would say we need to have a framework/guidelines on our side that follow best practices and include the knowledge of all participating companies. With the combined knowledge we can develop node images that are production-ready and follow all best practices. And have a good solution that provider-specific details can be implemented by everyone.